Hometask 8 - Http Client Module, Interceptor and Backend integration

Jonatan-Alvarez-EPAM opened the pull request:
Http header request with authorization token:

![token](https://user-images.githubusercontent.com/56736663/71461048-06ae5780-2774-11ea-837e-50ecf999f0d7.png)

aledpardo commented on src/app/app.module.ts:
```diff
@@ -15,9 +17,12 @@ import { AppComponentsRoutingModule } from './app-components/app-components-rout
   ],
   providers: [
     { provide: 'Storage', useValue: localStorage },
+    { provide: 'BASE_URL', useValue: 'http://localhost:3004' },
```
This is a great way to provide the URL! Very well done!

aledpardo commented on src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.html:
```diff
@@ -1,4 +1,4 @@
 <div class="rectangle">
     <span><a routerLink="/courses">Courses</a></span>
-    <span *ngIf="courseId">&nbsp;/&nbsp;{{courseTitle}}</span>
+    <span *ngIf="courseId">&nbsp;/&nbsp;{{courseName}}</span>
```
We may use CSS paddings to add spacing between HTML elements. For example:

```html
<span>
    <a routerLink="/courses">Courses</a>
</span>

<ng-container *ngIf="courseId">
    <span class='breadcrumb-separator'>/</span>
    <span>{{courseName}}</span>
</ng-container>
```

```css
.breadcrumb-separator{
    padding: 0 .5rem; // or 0 2px;
}
```

Jonatan-Alvarez-EPAM replied:
Done.

aledpardo commented on src/app/app-components/courses/add-course/add-course.component.ts:
```diff
@@ -42,18 +51,34 @@ export class AddCourseComponent implements OnInit {
+        length: this.courseInfo.length,
+        description: this.courseInfo.description,
+        isTopRated: this.courseInfo.isTopRated,
+      }).subscribe();
```
We can use the many `subscribe` callbacks to handle the result of the HTTP call. For example:

```typescript
...
.subscribe(
    (response) => { /* handle success */},
    (error) => { /* handle the error */ }
);
```

Jonatan-Alvarez-EPAM replied:
Done.

aledpardo commented on src/app/app-components/courses/courses-list/courses-list.component.ts:
```diff
@@ -12,30 +14,55 @@ import { CoursesService } from '@app/app-services';
   onLoadMore() {
-    console.log('[FAKE] Load more...');
+    const result$ = this.loadByPage(this.previousPage(), this.nextPage());
+    this.courses$ = zip(this.courses$, result$)
```
Well done using the `zip` operator!! 🥇

aledpardo commented on src/app/app-services/AuthInterceptor.ts:
```diff
@@ -0,0 +1,19 @@
+    constructor(
+        private readonly authService: AuthorizationService,
+        @Inject('Storage') private readonly localStorage: Storage) { }
+    intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
```
Just as a tip, here's how I usually format the code when a method has many arguments:

```typescript
constructor(
    private readonly authService: AuthorizationService,
    @Inject('Storage') private readonly localStorage: Storage
)  { }

intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    // code
}
```

Jonatan-Alvarez-EPAM replied:
Thanks for the tip, I was not sure about how to format all this params.

aledpardo commented on src/app/app-services/authorization.service.ts:
```diff
@@ -1,45 +1,54 @@
-  constructor(@Inject('Storage') private readonly localStorage: Storage) {
+  constructor(@Inject('Storage') private readonly localStorage: Storage,
+    @Inject('BASE_URL') private readonly BASE_URL: string,
+    private readonly httpClient: HttpClient) {
```
One more tip to format methods with many arguments:

```typescript
constructor(
    @Inject('Storage') private readonly localStorage: Storage,
    @Inject('BASE_URL') private readonly BASE_URL: string,
    private readonly httpClient: HttpClient
) {}
```

Jonatan-Alvarez-EPAM replied:
Ditto

Jonatan-Alvarez-EPAM commented on src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.ts:
```diff
@@ -11,16 +13,21 @@ import { CoursesService } from '@app/app-services';
     if (this.courseId) {
-      const course = this.coursesService.get(this.courseId);
-      this.courseTitle = course.title;
+      this.httpClient.get<Course>(`${this.BASE_URL}/courses`, {
```
Done.

Jonatan-Alvarez-EPAM commented:
Hi Ale,

Please feel free to check my second commit on this same branch. I addressed all your comments and fixed some minor issues I found.

Thanks,
Jon

aledpardo commented on src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.scss:
```diff
@@ -6,4 +6,9 @@
+}
+
+.course-name::before {
+    content: '/';
```
Nice solution!!

aledpardo commented on src/app/app-services/courses.service.ts:
```diff
@@ -1,83 +1,49 @@
-      return this.courses[index];
-    }
+  create(courseInfo: Course): Observable<Course> {
+    return this.httpClient.post<Course>(`${this.BASE_URL}/courses`, {
```
As the data types of the `create` method and the API are the same, we may simply pass to `post` the `courseInfo` object instead.

```typescript
return this.httpClient.post<Course>(`${this.BASE_URL}/courses`, courseInfo);
```

Jonatan-Alvarez-EPAM replied:
Done. (Thanks for catching this :smiley: )

aledpardo commented on src/app/app-services/courses.service.ts:
```diff
@@ -1,83 +1,49 @@
-    if (Number.isInteger(index)) {
-      this.courses.splice(index, 1);
-    }
+  update(courseInfo: Course): Observable<Course> {
```
What I said in the `create` method applies here as well.

Jonatan-Alvarez-EPAM replied:
Ditto

Jonatan-Alvarez-EPAM replied:
Ditto
